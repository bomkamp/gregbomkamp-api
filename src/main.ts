import { HttpApi } from '@aws-cdk/aws-apigatewayv2';
import { LambdaProxyIntegration } from '@aws-cdk/aws-apigatewayv2-integrations';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import {
  App,
  CfnOutput,
  CfnParameter,
  Construct,
  Duration,
  Stack,
  StackProps,
} from '@aws-cdk/core';

export class ApiLambdaStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const spotifyClientId = new CfnParameter(this, 'SPOTIFY_CLIENT_ID', {
      type: 'String',
      noEcho: true,
      minLength: 1,
    });

    const spotifyClientSecret = new CfnParameter(this, 'SPOTIFY_CLIENT_SECRET', {
      type: 'String',
      noEcho: true,
      minLength: 1,
    });

    const spotifyRefreshToken = new CfnParameter(this, 'SPOTIFY_REFRESH_TOKEN', {
      type: 'String',
      noEcho: true,
      minLength: 1,
    });

    const role: Role = new Role(this, 'LambdaIamRole', {
      roleName: this.stackName + '-lambda-role',
      description:
        'IAM Role used to allow Lambda execution & permit gateway access',
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
    });

    const gbApiFunction: NodejsFunction = new Function(this, 'GBApiFunction', {
      functionName: this.stackName + '-lambda',
      code: Code.fromAsset('dist'),
      handler: 'lambda.handler',
      runtime: Runtime.NODEJS_14_X,
      environment: {
        SPOTIFY_CLIENT_ID: spotifyClientId.valueAsString,
        SPOTIFY_CLIENT_SECRET: spotifyClientSecret.valueAsString,
        SPOTIFY_REFRESH_TOKEN: spotifyRefreshToken.valueAsString,
      },
      role,
      logRetentionRole: role,
      memorySize: 2048,
      timeout: Duration.seconds(15),
    });

    const apiGateway = new HttpApi(this, 'GBHttpApi', {
      apiName: this.stackName + '-http-api',
      corsPreflight: {
        allowOrigins: ['https://gregbomkamp.dev'],
      },
      defaultIntegration: new LambdaProxyIntegration({
        handler: gbApiFunction,
      }),
    });

    new CfnOutput(this, 'GBUrlOutput', {
      value: apiGateway.url ?? 'Error, no url created for gateway.',
    });
  }
}

const app = new App();

new ApiLambdaStack(app, 'my-stack-dev', {
  stackName: 'gregbomkamp-api',
  description: 'API used for fun integrations with Spotify/Instagram/IMDB/etc.',
});

app.synth();
