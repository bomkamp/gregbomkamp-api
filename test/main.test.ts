import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { ApiLambdaStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new ApiLambdaStack(app, 'test');

  expect(stack).not.toHaveResource('AWS::S3::Bucket');
  expect(stack).toHaveResource('AWS::Lambda::Function');
  expect(
    app.synth().getStackArtifact(stack.artifactId).template,
  ).toMatchSnapshot();
});
