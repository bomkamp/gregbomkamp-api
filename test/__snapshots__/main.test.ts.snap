// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "GBUrlOutput": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "GBHttpApi039092AB",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88ArtifactHash7A01D2CD": Object {
      "Description": "Artifact hash for asset \\"891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88\\"",
      "Type": "String",
    },
    "AssetParameters891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88S3BucketF6E13F9F": Object {
      "Description": "S3 bucket for asset \\"891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88\\"",
      "Type": "String",
    },
    "AssetParameters891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88S3VersionKeyFBEE480E": Object {
      "Description": "S3 key for asset version \\"891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88\\"",
      "Type": "String",
    },
    "SPOTIFYCLIENTID": Object {
      "MinLength": 1,
      "NoEcho": true,
      "Type": "String",
    },
    "SPOTIFYCLIENTSECRET": Object {
      "MinLength": 1,
      "NoEcho": true,
      "Type": "String",
    },
    "SPOTIFYREFRESHTOKEN": Object {
      "MinLength": 1,
      "NoEcho": true,
      "Type": "String",
    },
  },
  "Resources": Object {
    "GBApiFunction8A0EAFAB": Object {
      "DependsOn": Array [
        "LambdaIamRoleD8B84737",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88S3BucketF6E13F9F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88S3VersionKeyFBEE480E",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters891293fb00d2c34bae312392bf465e5741ce5897aac2fb8249435a2380d74b88S3VersionKeyFBEE480E",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "SPOTIFY_CLIENT_ID": Object {
              "Ref": "SPOTIFYCLIENTID",
            },
            "SPOTIFY_CLIENT_SECRET": Object {
              "Ref": "SPOTIFYCLIENTSECRET",
            },
            "SPOTIFY_REFRESH_TOKEN": Object {
              "Ref": "SPOTIFYREFRESHTOKEN",
            },
          },
        },
        "FunctionName": "test-lambda",
        "Handler": "lambda.handler",
        "MemorySize": 2048,
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaIamRoleD8B84737",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "GBHttpApi039092AB": Object {
      "Properties": Object {
        "CorsConfiguration": Object {
          "AllowOrigins": Array [
            "https://gregbomkamp.dev",
          ],
        },
        "Name": "test-http-api",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "GBHttpApiDefaultRoute7CA76E3B": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GBHttpApi039092AB",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$default",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "GBHttpApiDefaultRouteHttpIntegration1ef020c6ad87d81aa7a9f59049fe266a76B14458",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "GBHttpApiDefaultRouteHttpIntegration1ef020c6ad87d81aa7a9f59049fe266a76B14458": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GBHttpApi039092AB",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "GBApiFunction8A0EAFAB",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "GBHttpApiDefaultRoutetestGBHttpApiDefaultRouteA1154D08Permission83273C12": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GBApiFunction8A0EAFAB",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GBHttpApi039092AB",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GBHttpApiDefaultStage2C73E5DB": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "GBHttpApi039092AB",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "LambdaIamRoleD8B84737": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "IAM Role used to allow Lambda execution & permit gateway access",
        "RoleName": "test-lambda-role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
